hugo/content/posts/nix-bazel-crosspiling.md

```diff
@@ -0,0 +1,129 @@
+
+# Nix
+
+Nix is the package manager providing all the tools and SDKs needed to crosspile code. To pull this off, my [Nix Flake](https://github.com/ConsumingChaos/consumingchaos.github.io/tree/master/examples/nix-bazel-crosspiling/flake.nix) exposes two things, a `devShell` (making tools available in the terminal and VSCode) and a set `bazel` that contains derivations that get exposed to Bazel via [rules_nixpkgs](https://github.com/tweag/rules_nixpkgs). Keeping everything in the same Nix Flake lets me ensure there is only one set of tools and SDKs and that they're shared between the developer environment and Bazel.
+
+The main derivation is `bazel.config` which generates `config.bzl`, a Bazel file containing all of the Nix store paths I need exposed to Bazel. I opted to use absolute paths (where possible) for Nix provided packages because they are already content addressed, so any change to them will result in a new path, which in turn will invalidate Bazel's cache. Unfortunately this won't work with Bazel remote execution, but `rules_nixpkgs` doesn't support remote execution presently anyways. On the flip side, this skips the need for Bazel to make a giant symlink forest for each package, and in the case of some of the SDKs, works around the issue where they contain recursive symlinks, which Bazel chokes on.
+
+## SDKs
+
+Below are my rough instructions for acquiring the SDKs to each platform targeted in this post. The Apple and Microsoft ones are not publicly hostable, so it's left up to the reader to solve making them privately available for their project.
+
+**aarch64-unknown-linux-gnu**
+
+Downloaded from Chromium project. Reverse engineered from https://chromium.googlesource.com/chromium/src/build/+/refs/heads/main/linux/sysroot_scripts/install-sysroot.py
+
+1. Get `Sha1Sum` for `debian_{version}_arm64_sysroot.tar` from https://chromium.googlesource.com/chromium/src/build/+/refs/heads/main/linux/sysroot_scripts/sysroots.json
+2. Fill in URL `https://commondatastorage.googleapis.com/chrome-linux-sysroot/toolchain/`
+   `{Sha1Sum}/debian_{version}_arm64_sysroot.tar.xz`
+
+**universal-apple-darwin**
+
+Pulled from Xcode installation on a Mac.
+
+1. Run `(export OUTPUT=${PWD}/sdks/universal-apple-darwin/{version}.tar.xz`
+   `&& cd $(xcode-select -p)/Platforms/MacOSX.platform/Developer/SDKs/MacOSX{version}.sdk`
+   `&& tar -cvJf ${OUTPUT} .)`
+
+**universal-apple-ios**
+
+Pulled from Xcode installation on a Mac.
+
+1. Run `(export OUTPUT=${PWD}/sdks/universal-apple-ios/{version}.tar.xz`
+   `&& cd $(xcode-select -p)/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS{version}.sdk`
+   `&& tar -cvJf ${OUTPUT} .)`
+
+**universal-linux-android**
+
+Android NDK from https://developer.android.com/ndk/downloads
+
+**wasm32-unknown-unknown**
+
+By definition this shouldn't really require an SDK, but it does require `libclang_rt.builtins-*` which exist as a prebuilt in the `wasm32-wasi` SDK.
```

Comment:
I don't really understand why this one is here. Right below, you provide the same target and link. If you really need to specifically call out wasm32-unknown-unknown, I would do it right below the wasm32-wasi target and indicate the files come from the above target.